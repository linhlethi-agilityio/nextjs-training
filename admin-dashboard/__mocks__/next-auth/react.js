export const useSession = () => ({
  data: {
    user: {
      token: 'access-token',
      email: 'user@gmail.com',
      id: '1',
      name: 'User',
      role: 'admin',
    },
    expires: Date.now().toString(),
  },
  status: 'authenticated',
  update: async () => null,
});
