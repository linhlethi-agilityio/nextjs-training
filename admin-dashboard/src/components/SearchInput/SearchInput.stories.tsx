import type { Meta, StoryObj } from '@storybook/react';

// Components
import SearchInput from '.';

const meta: Meta<typeof SearchInput> = {
  component: SearchInput,
};

export default meta;
type Story = StoryObj<typeof SearchInput>;

export const Default: Story = {
  args: {
    placeholder: 'Search',
    onChange: () => null,
  },
};
