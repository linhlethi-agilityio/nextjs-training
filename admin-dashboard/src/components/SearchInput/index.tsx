'use client';

import { ChangeEvent, memo } from 'react';
import {
  Input,
  InputProps,
  InputGroup,
  InputRightElement,
  InputLeftElement,
} from '@chakra-ui/react';
import { useDebouncedCallback } from 'use-debounce';

// Icons
import { SearchIcon, SlidersIcon } from '@/icons';

interface SearchInputProps extends InputProps {
  placeholder?: string;
  defaultValue?: string;
  onSearch: (value: string) => void;
}

const SearchInput = ({
  placeholder = 'Search',
  defaultValue = '',
  onSearch,
  ...rest
}: SearchInputProps) => {
  const handleChange = useDebouncedCallback(
    (event: ChangeEvent<HTMLInputElement>) => {
      const value = event.target.value;

      onSearch(value);
    },
    300,
  );

  return (
    <InputGroup width="auto">
      <InputLeftElement pl={30} style={{ cursor: 'pointer' }}>
        <SearchIcon />
      </InputLeftElement>
      <Input
        bgColor="backgroundLight"
        pl={66}
        placeholder={placeholder}
        defaultValue={defaultValue}
        onChange={handleChange}
        {...rest}
      />
      <InputRightElement pr={10} style={{ cursor: 'pointer' }}>
        <SlidersIcon />
      </InputRightElement>
    </InputGroup>
  );
};

export default memo(SearchInput);
