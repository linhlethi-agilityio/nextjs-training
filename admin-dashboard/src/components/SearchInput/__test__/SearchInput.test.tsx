import { fireEvent, render } from '@testing-library/react';
import '@testing-library/jest-dom';

// Components
import SearchInput from '..';

const mockOnChange = jest.fn();

const mockProps = {
  onChange: mockOnChange,
};
describe('SearchInput component', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.runOnlyPendingTimers();
    jest.useRealTimers();
  });

  test('should match snapshot for SearchInput', () => {
    const { container } = render(<SearchInput {...mockProps} />);

    expect(container).toMatchSnapshot();
  });

  it('calls onChange with the deferred query value', () => {
    const handleChange = jest.fn();
    const { getByPlaceholderText } = render(
      <SearchInput placeholder="Search" value="" onChange={handleChange} />,
    );

    const inputElement = getByPlaceholderText('Search');
    fireEvent.change(inputElement, { target: { value: 'apple' } });

    // Simulate useDeferredValue delay
    jest.runAllTimers();

    expect(handleChange).toHaveBeenCalledWith({
      target: { value: 'apple' },
    });
  });
});
