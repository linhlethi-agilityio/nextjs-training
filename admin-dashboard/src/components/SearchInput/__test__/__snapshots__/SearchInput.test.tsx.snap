// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInput component should match snapshot for SearchInput 1`] = `
<div>
  <div
    class="chakra-input__group css-ck76jc"
    data-group="true"
  >
    <div
      class="chakra-input__left-element css-2lnh3j"
      style="cursor: pointer;"
    >
      <svg
        class="chakra-icon css-nyt2fo"
        focusable="false"
        viewBox="0 0 17 16"
      >
        <g
          clip-path="url(#clip0_16_102)"
        >
          <path
            d="M12.0725 10.344C13.0408 9.0227 13.4745 7.3845 13.2869 5.75716C13.0992 4.12982 12.304 2.63335 11.0603 1.56714C9.81666 0.50093 8.2163 -0.0563875 6.57938 0.00668566C4.94247 0.0697588 3.38973 0.748571 2.23182 1.90732C1.0739 3.06606 0.396199 4.61929 0.334297 6.25624C0.272396 7.8932 0.830858 9.49317 1.89796 10.7361C2.96505 11.9789 4.46209 12.7731 6.08957 12.9596C7.71704 13.1461 9.35493 12.7112 10.6755 11.742H10.6745C10.7045 11.782 10.7365 11.82 10.7725 11.857L14.6225 15.707C14.8101 15.8946 15.0644 16.0001 15.3297 16.0002C15.595 16.0003 15.8494 15.895 16.037 15.7075C16.2247 15.52 16.3302 15.2656 16.3303 15.0004C16.3303 14.7351 16.2251 14.4806 16.0375 14.293L12.1875 10.443C12.1518 10.4068 12.1134 10.3734 12.0725 10.343V10.344ZM12.3305 6.50001C12.3305 7.22228 12.1883 7.93748 11.9119 8.60477C11.6355 9.27206 11.2304 9.87837 10.7196 10.3891C10.2089 10.8998 9.6026 11.3049 8.93531 11.5813C8.26802 11.8577 7.55282 12 6.83055 12C6.10828 12 5.39308 11.8577 4.72579 11.5813C4.0585 11.3049 3.45218 10.8998 2.94146 10.3891C2.43074 9.87837 2.02561 9.27206 1.74921 8.60477C1.47281 7.93748 1.33055 7.22228 1.33055 6.50001C1.33055 5.04132 1.91001 3.64237 2.94146 2.61092C3.97291 1.57947 5.37186 1.00001 6.83055 1.00001C8.28924 1.00001 9.68819 1.57947 10.7196 2.61092C11.7511 3.64237 12.3305 5.04132 12.3305 6.50001Z"
            fill="currentColor"
          />
        </g>
        <defs>
          <clippath
            id="clip0_16_102"
          >
            <rect
              fill="white"
              height="16"
              transform="translate(0.330566)"
              width="16"
            />
          </clippath>
        </defs>
      </svg>
    </div>
    <input
      class="chakra-input css-kgmljx"
      placeholder="Search"
      value=""
    />
    <div
      class="chakra-input__right-element css-czbnbi"
      style="cursor: pointer;"
    >
      <svg
        class="chakra-icon css-nyt2fo"
        focusable="false"
        viewBox="0 0 17 16"
      >
        <g
          clip-path="url(#clip0_16_105)"
        >
          <path
            clip-rule="evenodd"
            d="M11.8306 2C11.6336 2 11.4385 2.0388 11.2565 2.11418C11.0746 2.18957 10.9092 2.30006 10.7699 2.43934C10.6306 2.57863 10.5201 2.74399 10.4447 2.92598C10.3694 3.10797 10.3306 3.30302 10.3306 3.5C10.3306 3.69699 10.3694 3.89204 10.4447 4.07403C10.5201 4.25602 10.6306 4.42138 10.7699 4.56066C10.9092 4.69995 11.0746 4.81044 11.2565 4.88582C11.4385 4.96121 11.6336 5 11.8306 5C12.2284 5 12.6099 4.84197 12.8912 4.56066C13.1725 4.27936 13.3306 3.89783 13.3306 3.5C13.3306 3.10218 13.1725 2.72065 12.8912 2.43934C12.6099 2.15804 12.2284 2 11.8306 2ZM9.38057 3C9.49532 2.43486 9.80193 1.92677 10.2484 1.56181C10.6949 1.19686 11.2539 0.997498 11.8306 0.997498C12.4072 0.997498 12.9662 1.19686 13.4127 1.56181C13.8592 1.92677 14.1658 2.43486 14.2806 3H16.3306V4H14.2806C14.1658 4.56515 13.8592 5.07324 13.4127 5.43819C12.9662 5.80314 12.4072 6.00251 11.8306 6.00251C11.2539 6.00251 10.6949 5.80314 10.2484 5.43819C9.80193 5.07324 9.49532 4.56515 9.38057 4H0.330566V3H9.38057ZM4.83057 7C4.43274 7 4.05121 7.15804 3.76991 7.43934C3.4886 7.72065 3.33057 8.10218 3.33057 8.5C3.33057 8.89783 3.4886 9.27936 3.76991 9.56066C4.05121 9.84197 4.43274 10 4.83057 10C5.22839 10 5.60992 9.84197 5.89123 9.56066C6.17253 9.27936 6.33057 8.89783 6.33057 8.5C6.33057 8.10218 6.17253 7.72065 5.89123 7.43934C5.60992 7.15804 5.22839 7 4.83057 7ZM2.38057 8C2.49532 7.43486 2.80193 6.92677 3.24844 6.56181C3.69494 6.19686 4.25389 5.9975 4.83057 5.9975C5.40724 5.9975 5.96619 6.19686 6.4127 6.56181C6.8592 6.92677 7.16581 7.43486 7.28057 8H16.3306V9H7.28057C7.16581 9.56515 6.8592 10.0732 6.4127 10.4382C5.96619 10.8031 5.40724 11.0025 4.83057 11.0025C4.25389 11.0025 3.69494 10.8031 3.24844 10.4382C2.80193 10.0732 2.49532 9.56515 2.38057 9H0.330566V8H2.38057ZM11.8306 12C11.4327 12 11.0512 12.158 10.7699 12.4393C10.4886 12.7206 10.3306 13.1022 10.3306 13.5C10.3306 13.8978 10.4886 14.2794 10.7699 14.5607C11.0512 14.842 11.4327 15 11.8306 15C12.2284 15 12.6099 14.842 12.8912 14.5607C13.1725 14.2794 13.3306 13.8978 13.3306 13.5C13.3306 13.1022 13.1725 12.7206 12.8912 12.4393C12.6099 12.158 12.2284 12 11.8306 12ZM9.38057 13C9.49532 12.4349 9.80193 11.9268 10.2484 11.5618C10.6949 11.1969 11.2539 10.9975 11.8306 10.9975C12.4072 10.9975 12.9662 11.1969 13.4127 11.5618C13.8592 11.9268 14.1658 12.4349 14.2806 13H16.3306V14H14.2806C14.1658 14.5651 13.8592 15.0732 13.4127 15.4382C12.9662 15.8031 12.4072 16.0025 11.8306 16.0025C11.2539 16.0025 10.6949 15.8031 10.2484 15.4382C9.80193 15.0732 9.49532 14.5651 9.38057 14H0.330566V13H9.38057Z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </g>
        <defs>
          <clippath
            id="clip0_16_105"
          >
            <rect
              fill="white"
              height="16"
              transform="translate(0.330566)"
              width="16"
            />
          </clippath>
        </defs>
      </svg>
    </div>
  </div>
</div>
`;
